File: ch2/Insertion_Sort/algorithm.cs
using System;

public class algorithm 
{
    public static T[] insertion_sort<T>(T[] List) where T:IComparable<T>
    {
        for(int i = 1 ; i < List.Length ; i++) {
            T current = List[i];
            int position = i-1;
            while ((position >= 0) && (current.CompareTo(List[position])<0)){
                List[position+1] = List[position];
                position = position - 1;
            }
            List[position+1] = current;
        }
        return List;
    }

    public static void Main(string[] args) {
        int[] numbers = {1223,1,23,54,655,23,65};
        int[] sorted = algorithm.insertion_sort(numbers);
        Console.WriteLine("[{0}]", string.Join(", ", sorted));

    }
}

File: ch2/Insertion_Sort/algorithm.go
package main

import "fmt"

func insertion_sort(List []int) []int {
	for i := 1; i < len(List); i++ {
		current := List[i]
		position := i - 1

		for position >= 0 && current < List[position] {
			List[position+1] = List[position]
			position -= 1
		}
		List[position+1] = current
	}
	return List
}

func main() {
	List := []int{5, 4, 1, 7, 4, 9, 10, 23, 12, 6, 3, 1}
	fmt.Println(insertion_sort(List))
}
